Backport 4.x

boesing opened the pull request.

boesing committed:
spelling fixes

boesing committed:
Dropped PHP 5.6 and PHP 7.0

Updated packages in composer to latest stable version and PHP requirements
changed to PHP ^7.1
Updated suggestion in composer.

boesing committed:
Changed self:: to $this-> for expectException* methods

These methods are not static.

boesing committed:
Use self:: instead of $this-> for assert* methods

boesing committed:
Updated CONTRIBUTING.md

boesing committed:
Optimize more PHP function calls and issets

boesing committed:
Alphabetical order of use statements

boesing committed:
Fix problem with string callables failing in PHP 5.6.

boesing committed:
Adds CHANGELOG for #206

boesing committed:
3.3.1 readiness

Conflicts:
	CHANGELOG.md

boesing committed:
#216 checking that the `AbstractFactory` is called with a resolved alias name, and not the alias itself

boesing committed:
Started optimizing abstractFactories.

boesing committed:
Removed obsolete test setup items.

boesing committed:
Merge branch 'feature/221' into develop

Close #221

boesing committed:
Merge branch 'feature/230' into develop

Close #230

boesing committed:
Use default values for type-hinted arguments if type not found in container

Per #239, currently when an argument is type-hinted but has a default
value (typically `null`), the factory raises an exception about not
being able to retrieve the value from the container. In such cases, we
can use the default value for the argument during injection.

boesing committed:
Merge branch 'hotfix/239-reflection-factory-default-values'

Close #243
Fixes #239

boesing committed:
Merge branch 'hotfix/233-parameter-type-hints' into develop

Forward port #233

Also brings in changes specific to develop branch.

boesing committed:
Merge pull request #234 from malukenho/enhancement/use-imported-function

Avoid usage of leading `\` on imported functions

boesing committed:
Merge pull request #237 from fhein/TEST-Stability-Over-Internal-State-Changes

Adds test for issue #236

boesing committed:
#241 corrected CS in newly introduced tests, removed `equalsTo` for strings

boesing committed:
Removed unused class import

boesing committed:
Updated to use laminas-container-config-test library

All these tests implemented  here previously for shared services are moved
into separate library so it should be used instead.

boesing committed:
ReflectionBasedAbstractFactory should not be expected to instantiate private constructors

boesing committed:
Update annotations and comments for consistency

boesing committed:
Merge pull request #35 from pine3ree/exception-messages

update exception messages

boesing committed:
Merge pull request #38 from samsonasik/fix-travis-build

Fix travis build : cs fixes

boesing committed:
fix: alphabetize import statements in generated factories

Previously, the imports were _mostly_ alphabetized. However, with the
switch to Laminas, the container import was now happening incorrectly
with regards to the factory interface. Additionally, it makes sense to
alphabetize to ensure we follow our own CS.

This patch updates the factory generation code to alphabetize import
statements before injecting them in the template.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

boesing committed:
update "mikey179/vfsstream" to ^1.6.8

Signed-off-by: Abdul Malik Ikhsan <samsonasik@gmail.com>

boesing committed:
docs: creates CHANGELOG entry for #40

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

boesing committed:
Updated min php version to 7.2 and supported versions to 7.4

Signed-off-by: Gary Hockin <gary@hock.in>

boesing committed:
Updated min php version to 7.3 and supported versions to 7.4

Signed-off-by: Gary Hockin <gary@hock.in>

boesing committed:
Fixed conflict in zend-code

Signed-off-by: Gary Hockin <gary@hock.in>

boesing committed:
tidied code up

Signed-off-by: Gary Hockin <gary@hock.in>

boesing committed:
Merge pull request #43 from GeeH/geeh/#3-hotfix-from-old-repo

boesing committed:
added strict types to tests and factory stub

Signed-off-by: Gary Hockin <gary@hock.in>

boesing committed:
Update docs deployment workflow

Signed-off-by: Geert Eltink <dev@elt.ink>

boesing committed:
Fix changelog next version

Signed-off-by: Geert Eltink <dev@elt.ink>

boesing committed:
add phpunit9 support

Signed-off-by: gennadigennadigennadi <gennadijanzen@icloud.com>

boesing committed:
add use statements for Prophecy\PhpUnit\ProphecyTrait

Signed-off-by: gennadigennadigennadi <gennadijanzen@icloud.com>

boesing committed:
replace BackportAssertionsTrait

Signed-off-by: gennadigennadigennadi <gennadijanzen@icloud.com>

boesing committed:
Ignore platform requiments to install deps on php8

Signed-off-by: Witold Wasiczko <witold@wasiczko.pl>

boesing committed:
Fix lowest in php 8.0

Signed-off-by: Witold Wasiczko <witold@wasiczko.pl>

boesing committed:
Add `~8.0.0` constraint

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
Merge pull request #51 from snapshotpl/pass-php8

Initial support for PHP 8.0

boesing committed:
qa: add phpunit result cache to `.gitignore`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: use `nightly` for travis as they still did not come up with `8.0`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: ensure composer does not run on v2

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: do not ignore platform reqs when installing dependencies

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: fix backport regression

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: use dependency-plugin as dev-dependency to force `ocramius/proxy-manager` to use `laminas` equivalent of `zend-code`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: allow plugins so the dependency-plugin can do its thing

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: add conflict to `laminas-code` less than 3.4 as it does not have 7.4 compatibility

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: add conflict with `zendframework/zend-code` less than `3.3.1` aswell as `laminas/laminas-code` less than `3.3.1`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

snapshotpl commented on .travis.yml:
```diff
@@ -6,45 +6,43 @@ cache:
-        - CS_CHECK=true
-        - TEST_COVERAGE=true
-    - php: 7.2
+    - php: nightly
```
Travis support `8.0` now

boesing replied:
Since when? Tested that 3 days ago and it did not.
Any official resource to verify your statement before I start changing `.travis.yml` twice?

boesing replied:
https://travis-ci.community/t/php-8-0-missing/10132/23

snapshotpl commented on .travis.yml:
```diff
@@ -6,45 +6,43 @@ cache:
       env:
         - DEPS=lowest
-    - php: 7.2
+        - COMPOSER_ARGS="$COMPOSER_ARGS --ignore-platform-reqs"
```
Are we ok with `ignore-platform-reqs` ? I think it's risky to use something that is not right tested. Which dependencies require this flag?

boesing replied:
Since 3rd-party dependencies do not provide proper constraints, this is still needed.
It is risky but thats how it is for now and thats how it is in almost every other laminas/mezzio package due to the lack of support in other libs.

snapshotpl commented on .travis.yml:
```diff
@@ -6,45 +6,43 @@ cache:
 before_install:
   - if [[ $TEST_COVERAGE != 'true' ]]; then phpenv config-rm xdebug.ini || return 0 ; fi
+  # Restore composer back to v1
+  - composer self-update --1
```
Why?

boesing replied:
Due to incompatibility with `ocramius/package-versions` and I dont want to spend too much time in this to fix as it works with composer v1 without any issue.

https://travis-ci.com/github/laminas/laminas-servicemanager/jobs/452925196#L456

boesing committed:
qa: remove `DEPS=latest` as we are always supporting latest dependencies

`composer.lock` is not committed anymore, thus, there is no need for `latest` and `composer update` anymore

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: dev optimizations to ensure build pipeline can run without "hacks"

- require `composer/package-versions-deprecated` in order to get proper installation on PHP 7.3 and 8.0
- require `phpbench/phpbench` alpha as it supports PHP 8.0
- only ignore platform-requirement for PHP until `composer/proxy-manager` supports PHP 8.0

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: add native PHP 8.0 support for travis and use `nightly` with `allow_failures`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

froschdesign commented on composer.json:
```diff
@@ -23,33 +23,35 @@
     "suggest": {
-        "laminas/laminas-stdlib": "laminas-stdlib ^2.5 if you wish to use the MergeReplaceKey or MergeRemoveKey features in Config instances",
-        "ocramius/proxy-manager": "ProxyManager 1.* to handle lazy initialization of services"
+        "laminas/laminas-stdlib": "laminas-stdlib ^2.7.7 | ^3.1 if you wish to use the MergeReplaceKey or MergeRemoveKey features in Config instances",
```
The "require" section already contains laminas-stdlib therefore this line can be removed.

weierophinney commented on src/AbstractFactory/ConfigAbstractFactory.php:
```diff
@@ -64,8 +72,8 @@ public function __invoke(\Interop\Container\ContainerInterface $container, $requ
 
-        if ($serviceDependencies !== array_values(array_map('strval', $serviceDependencies))) {
-            $problem = json_encode(array_map('gettype', $serviceDependencies));
+        if ($serviceDependencies !== array_values(array_map('\strval', $serviceDependencies))) {
```
Functions and classes referenced inside of strings are always considered globally qualified; there's no reason to prefix them.

weierophinney commented on src/Exception/ContainerModificationsNotAllowedException.php:
```diff
@@ -10,9 +12,24 @@
+     * @param string $service Name of service that already exists.
+     * @return self
+     */
+    public static function fromExistingService($service)
```
Since this is a new method, and we're updating to PHP 7.4 as the minimum supported version, provide explicit argument and return typehints here, please.

weierophinney commented on src/Exception/CyclicAliasException.php:
```diff
@@ -8,11 +10,41 @@
+     * @param string[] $aliases map of referenced services, indexed by alias name (string)
+     * @return self
+     */
+    public static function fromCyclicAlias($alias, array $aliases)
```
Provide all argument and return type hints on new methods, please.

weierophinney commented on src/Exception/InvalidArgumentException.php:
```diff
@@ -9,10 +11,44 @@
+     * @param mixed $initializer
+     * @return self
+     */
+    public static function fromInvalidInitializer($initializer)
+    {
+        return new self(sprintf(
+            'An invalid initializer was registered. Expected a callable or an'
+            . ' instance of "%s"; received "%s"',
+            InitializerInterface::class,
+            is_object($initializer) ? get_class($initializer) : gettype($initializer)
+        ));
+    }
+
+    /**
+     * @param mixed $abstractFactory
+     * @return self
+     */
+    public static function fromInvalidAbstractFactory($abstractFactory)
```
Typehints, please.

weierophinney commented on src/Tool/FactoryCreator.php:
```diff
@@ -39,6 +56,11 @@ public function __invoke(ContainerInterface $container, $requestedName, array $o
 
 EOT;
```
Since we're bumping the minimum supported PHP version... PHP 7.3+ allow you to indent heredocs and nowdocs. The closing marker (in this case, `EOT;`) defines the amount of indentation that will be stripped from the start of each line of the heredoc/nowdoc. I've done this in a few other components, and it really helps improve readability.

weierophinney commented on src/Tool/FactoryCreator.php:
```diff
@@ -110,7 +135,8 @@ function (ReflectionParameter $argument) {
         }
 
         return array_map(function (ReflectionParameter $parameter) {
```
This should have a return typehint of `?string`.

weierophinney commented on src/Tool/FactoryCreator.php:
```diff
@@ -140,4 +166,13 @@ private function createArgumentString($className)
+    {
+        $imports = array_merge(self::IMPORT_ALWAYS, [$className]);
+        sort($imports);
+        return implode("\n", array_map(function ($import) {
```
This should have a return typehint of `string`.

weierophinney commented on test/AbstractFactory/ConfigAbstractFactoryTest.php:
```diff
@@ -185,8 +187,8 @@ public function testExceptsWhenConfigNotSet()
         $serviceManager = new ServiceManager();
-        self::expectException(ServiceNotCreatedException::class);
-        self::expectExceptionMessage('Cannot find a config array in the container');
+        $this->expectException(ServiceNotCreatedException::class);
+        $this->expectExceptionMessage('Cannot find a config array in the container');
```
Arggghhh!

We need to likely decide on which syntax we use, so that we don't have commits like this where we flip between styles.

Prior to the last year or so, we were consistently using instance access; in the past year, particularly with some of the updates to add Psalm support, there have been a number of switches to static access. We need to pick one and be consistent everywhere.

For this patch, I'd suggest leaving these alone, as they add a lot of extra noise to the patch.

weierophinney commented on test/Exception/CyclicAliasExceptionTest.php:
```diff
@@ -16,13 +18,103 @@
+    public function cyclicAliasProvider()
+    {
+        return [
+            [
```
Please give these names (e.g. "a -> a", "a -> b -> a", etc.).

weierophinney commented on test/Exception/CyclicAliasExceptionTest.php:
```diff
@@ -16,13 +18,103 @@
+     *
+     * @return string[][]|string[][][]
+     */
+    public function cyclicAliasProvider()
```
Return typehint, please.

weierophinney commented on test/Exception/CyclicAliasExceptionTest.php:
```diff
@@ -16,13 +18,103 @@
      * @param string[] $aliases
      * @param string   $expectedMessage
+     */
+    public function testFromCyclicAlias($alias, array $aliases, $expectedMessage)
```
Typehints, please.

weierophinney reviewed (changes requested):
Main comments:

- We're bumping the minimum supported PHP version to 7.4. So all NEW methods should have full typehinting in place; this includes closures.
- Add indenting to heredoc/nowdoc definitions (I have a comment on that later).
- Don't switch between `$this->` and `self::` in the tests; it's a lot of noise, and we really need to have a discussion about this in the TSC to decide which one we're going to use going forward. For now, instance usage is consistent with other components, and reduces the size of this patch considerably.

weierophinney committed:
qa: incorporate feedback from reviews

- Add typehints for any new function/method declarations.
- Do not use FQ class names; functions in strings do not need qualification.
- Indent heredoc/nowdoc declarations.
- Provide typehints on closure definitions.
- Do not use `self::` in unit tests.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: remove suggestion for package that is now required

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: remove nightly targets from travis workflow

We're pinned to `~8.0.0`, and nightly targets 8.1.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: set alternative minimum for ocramius/proxy-manager under PHP 8

In order to get laminas-code versions that work under PHP 8, we need to have an ocramius/proxy-manager version that no longer uses zend-code.
That version is 2.11.0.
Also, we need to explicitly select a newer version of doctrine/annotations to fix conflicts with phpbench.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

boesing committed:
qa: add `CHANGELOG`entry

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: apply feedback from reviews

- remove duplicated test
- adds typehints

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>
